src/types/mod.rs

```diff
@@ -424,6 +424,64 @@ impl ValueInternal for JsNumber {
     }
 }
 
+/// A JavaScript Date object.
+#[repr(C)]
+#[derive(Clone, Copy)]
+#[cfg(feature = "napi-runtime")]
+pub struct JsDate(raw::Local);
+
+#[cfg(feature = "napi-runtime")]
+impl Value for JsDate { }
+
+#[cfg(feature = "napi-runtime")]
+impl Managed for JsDate {
+    fn to_raw(self) -> raw::Local { self.0 }
+
+    fn from_raw(_: Env, h: raw::Local) -> Self { JsDate(h) }
+}
+
+#[cfg(feature = "napi-runtime")]
+impl JsDate {
+    pub const MIN_VALID_VALUE: f64 = -8640000000000000.0;
+    pub const MAX_VALID_VALUE: f64 = 8640000000000000.0;
+
+    pub(crate) fn new_internal<'a>(env: Env, value: f64) -> Handle<'a, JsDate> {
+        unsafe {
+            let mut local: raw::Local = std::mem::zeroed();
+            neon_runtime::date::new_date( env.to_raw(), &mut local, value);
+            Handle::new_internal(JsDate(local))
+        }
+    }
+
+    pub fn new<'a, C: Context<'a>, T: Into<f64>>(cx: &mut C, time: T) -> Handle<'a, JsDate> {
```

Comment:
I think this is an interesting idea. I'm wondering if its consistent with existing notions of when we return `Result`.

Reply:
@kjvalencik Interesting idea. It feels a _little_ unidiomatic to me, or somehow strange to say "I failed to create a `Date` and btw here it is."

Oh interesting -- in the RFC, I had it returning a `JsResult` in case the `napi_status` is not `napi_ok`, but is that impossible? I guess this API can never throw a JS exception?

If that's the case, another option would be two or three different entry points, with a custom `DateErrorKind`:

```rust
pub enum DateErrorKind {
    Overflow,
    Underflow,
}

impl JsDate {
    // checks the range and returns an Error if out of range
    pub fn try_new<'a, C: Context<'a>, V: Into<f64>>(_: &mut S, value: V) -> Result<Handle<'a, JsDate>, DateErrorKind>;
    // checks the range and panics if out of range
    pub fn new<'a, C: Context<'a>, V: Into<f64>>(_: &mut S, value: V) -> Handle<'a, JsDate>;
    // doesn't check the range and produces a Date value no matter what
    pub fn new_unchecked<'a, C: Context<'a>, V: Into<f64>>(_: &mut S, value: V) -> Handle<'a, JsDate>;
}
```

And then for convenience we'd have `cx.date(v)` be equivalent to `JsDate::new(&mut cx, v)` and maybe `cx.try_date(v)` would be equivalent to `JsDate::try_new(&mut cx, v)` (similar to `cx.try_string(...)`). I wouldn't make a convenience method for `new_unchecked` since it's probably a rare case.

Reply:
I think the three different kinds of entrypoints might be too much API surface area. I favor going with the 'checks the range and panics if out of range' API and adding additional APIs if users find it insufficient for their use cases.

Reply:
I agree that three seems like one to many, but a panic on a recoverable error doesn't seem ideal to me. I _really_ like the `try_new` with the `DateError`. One of the reasons I like it is it allows us to extend this function to accept `i64` as well since out of range for the date implies overflow on an `f64`.

Reply:
+1. Will implement `try_new`